subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java

```diff
@@ -131,96 +124,58 @@ private void resizeTo(int newBuildProgressLabelCount) {
         for (int i = newBuildProgressLabelCount - 1; i >= previousBuildProgressLabelCount; --i) {
             unusedProgressLabels.push(progressArea.getBuildProgressLabels().get(i));
         }
-
-        // Try to empty the unassigned progress operations
-        while (!unusedProgressLabels.isEmpty() && !unassignedProgressOperations.isEmpty()) {
-            attach(unassignedProgressOperations.pop());
-        }
     }
 
     private void attach(ProgressOperation operation) {
-        // Skip attach if a child is already present or no progress message to display
-        if (isChildAssociationAlreadyExists(operation.getOperationId())) {
+        // Skip attach if a children is already present
+        if (!operation.getChildren().isEmpty() || !isRenderable(operation)) {
             return;
         }
 
-        AssociationLabel association = null;
-
         // Reuse parent label if possible
         if (operation.getParent() != null) {
-            addDirectChildOperationId(operation.getParent().getOperationId(), operation.getOperationId());
-            association = operationIdToAssignedLabels.remove(operation.getParent().getOperationId());
-            if (association != null) {
-                unusedProgressLabels.push(association.label);
-                association = null;
-            }
-        }
-
-        if (!isRenderable(operation)) {
-            return;
+            detach(operation.getParent().getOperationId());
         }
 
-        // No parent? Try to use a new label
-        if (!unusedProgressLabels.isEmpty()) {
-            association = new AssociationLabel(operation, unusedProgressLabels.pop());
+        // No more unused label? Try to resize.
+        if (unusedProgressLabels.isEmpty()) {
+            int newValue = operationIdToAssignedLabels.size() + 1;
+            resizeTo(newValue);
         }
 
-        if (association == null) {
-            unassignedProgressOperations.addLast(operation);
+        // Try to use a new label
+        if (unusedProgressLabels.isEmpty()) {
```

Comment:
:o: We're doing the same check as above. Can we just inline `resizeTo(oITAL.size() + 1)` here?